src/main/java/ui/components/pickers/MilestonePickerState.java

```diff
@@ -30,82 +34,244 @@ public MilestonePickerState(List<PickerMilestone> milestones, String userInput)
+                });
     }
 
     /**
-     * Toggles the milestone that has milestoneName as its exact name
-     * <p>
-     * This method is case-sensitive
+     * Toggles the selection status of the milestone that has milestoneName as its exact name
      *
+     * This method is case-sensitive
      * @param milestoneName
      */
     public final void toggleExactMatchMilestone(String milestoneName) {
-        Optional<PickerMilestone> onlyMatchingMilestone = getExactMatchMilestone(currentMilestonesList, milestoneName);
-        onlyMatchingMilestone.ifPresent(milestone -> toggleMilestone(milestone));
+        Optional<PickerMilestone> exactMatchMilestone = getExactMatchMilestone(allMilestones, milestoneName);
+        exactMatchMilestone.ifPresent(this::toggleMilestone);
     }
 
     /**
-     * Toggles the milestone that is the first match to the given query
-     * <p>
-     * This method is NOT case-sensitive
+     * Toggles the selection status of the first milestone in the list
      *
-     * @param milestoneList
+     * @param milestones
      */
-    private final void toggleFirstMatchingMilestone(List<PickerMilestone> milestoneList) {
-        Optional<PickerMilestone> firstMatchingMilestone = getFirstMatchingMilestone(milestoneList);
-        firstMatchingMilestone.ifPresent(this::toggleMilestone);
+    private final void toggleFirstMilestone(List<PickerMilestone> milestones) {
+        Optional<PickerMilestone> firstMilestone = milestones.stream()
+                                                            .findFirst();
+        firstMilestone.ifPresent(this::toggleMilestone);
     }
 
     /**
-     * Changes the selection statuses of all milestones in the milestone list,
-     * such that the new given milestone is toggled, and the rest are not selected
+     * Changes the selection statuses of all milestones in allMilestones,
+     * such that the given milestone is toggled, and the rest are not selected
      *
      * @param milestone
      */
     private void toggleMilestone(PickerMilestone milestone) {
-        this.currentMilestonesList.stream()
-                .forEach(listMilestone ->
-                        listMilestone.setSelected(listMilestone.equals(milestone) && !listMilestone.isSelected()));
+        allMilestones.stream()
+                .forEach(listMilestone -> listMilestone.setSelected(listMilestone.equals(milestone)
+                        && !listMilestone.isSelected()));
+    }
+
+    public List<PickerMilestone> getAllMilestones() {
+        return allMilestones;
+    }
+
+    public List<PickerMilestone> getBestMatchingMilestones() {
+        return bestMatchingMilestones;
+    }
+
+    /**
+     * Populates bestMatchingMilestones with milestones from allMilestones that best match
+     * {@code userInput}.
+     *
+     * The added milestones are references to the actual elements in allMilestones, and thus should not be
+     * unnecessarily mutated
+     * @param userInput
```

Comment:
Put explanation on param, what is `userInput`. I think it's query sentence? `userInput` is very general and might become ambiguous.